### PR TITLE
Close #11512: Add coloured usernames in multiplayer servers by group

### LINKED ISSUE #11512: Feature: Colored usernames on servers, depending on assigned role
**OS:** [Windows 10]
**Version:** [0.2.6]
**Commit/Build:** [06bbf5d]

Something that would be nice to have is the ability to assign a color to a role on multiplayer servers, and then display usernames from people of that role in the assigned color in the chat.

### PR DESCRIPTION
**Pull request by evmarcus**
#11512 

This feature changes the color of a user's username in the chat based on their role on multiplayer servers. See the image below for changes:
Screenshot 2024-04-11 203416

We didn't implement the fix like @IntelOrca suggested -- is this ok? How do you feel about the colors we chose?

### CODE REVIEW

**Comment by janisozaur** [comment]
Servers usually change groups



The solution you propose won't work properly for such servers.

**Comment by evmarcus** [comment]
The new solution we're working on automatically sets the colors of the username to the same colors in the name of their group, like this: 
image
image
Should we finalize this solution, or do you have another idea in mind?

**Comment by janisozaur** [comment]
This looks much better now

**Comment by evmarcus** [comment]
We've finished a working version of the new specification of the problem. Does this look good?

We scan the name of the group that the user is in. If there are no added colors to the group name, we just set the username to the default baby blue, but if there are colors, we extract those and use them to format the username in the chat.

**Comment by janisozaur** [comment]
I'll take a look at it this week

File: C [hunk 1]
```diff
@@ -0,0 +1 @@
+:
```
Comment (Harry-Hopkinson) [inline C:1]: This looks like an accidental change.

Comment (mattbilik) [reply C]: Fixed!

File: src/openrct2/network/NetworkBase.cpp [hunk 2]
```diff
@@ -715,11 +715,65 @@ const char* NetworkBase::FormatChat(NetworkPlayer* fromplayer, const char* text)
...
+            formatted += "{BABYBLUE}";
+            formatted += fromplayer->Name;
+        }
+        else
+        {
+            size_t j = 0;
+            for (size_t i = 0; i < colors.size(); ++i)
+            {
+                formatted += colors[i];
+                size_t numcharacters = (fromplayer->Name.size() / colors.size()) + j;
```
Comment (janisozaur) [inline src/openrct2/network/NetworkBase.cpp:764]: I don't expect player name or colors to change in each iteration, please move `fromplayer->Name.size() / colors.size()` calculation to before the loop.

**Review by janisozaur — APPROVED** [review]

File: src/openrct2/network/NetworkBase.cpp [hunk 3]
```diff
@@ -715,11 +715,66 @@ const char* NetworkBase::FormatChat(NetworkPlayer* fromplayer, const char* text)
 {
     static std::string formatted;
     formatted.clear();
-    formatted += "{OUTLINE}";
+
     if (fromplayer != nullptr)
```
Comment (Gymnasiast) [inline src/openrct2/network/NetworkBase.cpp:719]: Could you use variable names that are in `camelCase`?

- `fromPlayer`
- `groupName`
- `proportionalSize`
- `numCharacters`

Additionally, we use British English in our codebase, so please rename anything with `color` to `colour`,

File: src/openrct2/network/NetworkBase.cpp (hunk 2 above)
Comment (evmarcus) [reply src/openrct2/network/NetworkBase.cpp]: Moved!

File: src/openrct2/network/NetworkBase.cpp (hunk 3 above)
Comment (evmarcus) [reply src/openrct2/network/NetworkBase.cpp]: Cameled and coloured!

**Comment by janisozaur** [comment]
Needs changelog entry

**Review by Gymnasiast — APPROVED** [review]
Code looks good, want to take it for a brief test.

File: distribution/changelog.txt [hunk 4]
```diff
@@ -4,6 +4,7 @@
 - Improved: [#21769] Expose “animation is backwards” wall property in Tile Inspector.
 - Change: [#21715] [Plugin] Remove access to the internal `owner` property. Note: `ownership` is still accessible.
 - Fix: [#866] Boat Hire boats get stuck entering track.
+- Fix: [#11512] Coloured usernames by group on multiplayer servers.
```
Comment (Gymnasiast) [inline distribution/changelog.txt:7]: This is a feature, rather than a bugfix. (Keep in mind that features are placed higher in the changelog as well.)

Comment (evmarcus) [reply distribution/changelog.txt:7]: janisozaur told us on the Discord that "Fix" is fitting here because there is a prior issue filed that our pr solves

**Comment by Gymnasiast** [comment]
Test is also okay. Please update the changelog and I’ll merge it.

**Review by Gymnasiast — APPROVED** [review]

**Review by janisozaur — APPROVED** [review]

**Comment by janisozaur** [comment]
There is still an issue around this code: https://github.com/OpenRCT2/OpenRCT2/issues/21972

**Pull request merged**